rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,221 @@
+1) Local Health Checks - used to check this clusters IP addresses are responding for the expected host.
+2) Remote Health Checks - used to check that the other IP addresses in the zone file are responding for that zones host.
+
+When a local health check is failing, this IP will not be published to the zone.
+
+When a remote health check is failing a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
+
+If all local and remote health checks for a particular host are marked as unhealthy, all known IPs are published, to avoid an NXDOMAIN response.
+
+## OCM / ACM
+This is very similar to a single cluster, with the caveat that the "single" cluster is the hub cluster for this OCM. The primary difference is that the gateway will be a Kuadrant gatewayClass.
+
+### OCM / ACM Health Checks
+This is also the same as single cluster.
+
+## Migrating from Single cluster to Multiple cluster
+
+In the single cluster: 
+- Edit all relevant provider secrets and set:
+  - `ClusterID` to some unique value
+
+In the additional clusters:
+- Copy the DNS Policies and provider secrets from the single cluster to any additional clusters, and update the provider secrets with unique `ClusterID` values.
+- (optional) Create a gateway and workload which uses the same hosts as the single cluster.
+
+## Migrating from single/multiple clusters to OCM
+
+This is not covered.
+
+## Cleanup
+
+Cleanup of a cluster's zone file entries is triggered by deleting the DNS Policy that caused the DNS Records to be created.
+
+The DNS policy will have a finalizer, when it is deleted the controller will mark the created DNS Records for deletion.
+
+The DNS Records will have a finalizer from the DNS Operator which will clean the records from the zone before removing the finalizer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## General Logic Overview
+
+The general flow for DNS Records follows a single path in the Kuadrant operator, where it will always output a DNS Record which specifies everything needed for this workload on this cluster, and is unaware of the concept of leaderless DNS, or phrased more simply:
+Build a DNS Record (kuadrantRecord) based on the local gateways and DNS Policy.
+
+This kuadrantRecord is reconciled by the DNS Operator, where there are 2 possible routes:
+
+Where a clusterID is specified in the relevant ProviderSecret the DNS Operator will act with the DNS Provider's zone as the authority and ensure that the records in the zone relevant to it's cluster are accurate, when cleaning up a DNS Record in this scenario the operator will ensure all records for the local cluster are removed and then prune unresolvable records from the DNS Provider zone.
```

Comment:
```suggestion
Where a clusterID is specified in the relevant ProviderSecret the DNS Operator will act with the DNS Provider's zone as the authority and ensure that the records in the zone relevant to the hosts defined within gateways on it's cluster are accurate, when cleaning up a DNS Record in this scenario the operator will ensure all records for the local cluster are removed and then prune unresolvable records from the DNS Provider zone.
```

Reply:
yeah ok so from reading this it seems it will be fine to set a cluster id even for a single cluster?

Reply:
question would it be possible for the controller to identify a clash of clusterids and report it as an error?

Reply:
> question would it be possible for the controller to identify a clash of clusterids and report it as an error?

that's a pretty interesting question. I don't think so, as the A records with a local ID and different IPs could be the local cluster's old IP addresses that need to be updated.